File: ai/select-algorithm-python/src/compare_all.py
"""
Compare All Algorithms — Unified comparison runner.

Executes all 9 combinations (3 algorithms × 3 similarity metrics) in a single
invocation and prints a formatted comparison table.

Algorithms: IVF, HNSW, DiskANN
Metrics: COS, L2, IP
"""
import os
import time
from typing import Dict, List, Any

from tabulate import tabulate
from utils import (
    get_clients_passwordless, get_config, read_file_return_json,
    insert_data
)

# Index definitions: (algo_label, kind, extra_params)
ALGORITHMS = [
    ("IVF", "vector-ivf", {"numLists": 1}),
    ("HNSW", "vector-hnsw", {"m": 16, "efConstruction": 64}),
    ("DiskANN", "vector-diskann", {"maxDegree": 32, "lBuild": 50}),
]

METRICS = ["COS", "L2", "IP"]


def get_compare_config() -> Dict[str, Any]:
    """Load comparison-specific configuration from environment variables."""
    config = get_config()
    config["query_text"] = os.getenv("QUERY_TEXT", "luxury hotel near the beach")
    config["top_k"] = int(os.getenv("TOP_K", "5"))
    return config


def index_name(algo: str, metric: str) -> str:
    """Generate canonical index name: vector_{algo}_{metric}."""
    return f"vector_{algo.lower()}_{metric.lower()}"


def get_existing_index_names(collection) -> List[str]:
    """Return names of existing indexes on the collection."""
    return [idx["name"] for idx in collection.list_indexes()]


def drop_vector_indexes(collection, vector_field: str) -> None:
    """Drop all existing vector indexes on *vector_field*."""
    for idx in collection.list_indexes():
        name = idx.get("name", "")
        key = idx.get("key", {})
        if vector_field in key and key[vector_field] == "cosmosSearch":
            collection.drop_index(name)


def create_vector_index(collection, name: str, kind: str, vector_field: str,
                        dimensions: int, similarity: str,
                        extra_params: Dict[str, Any]) -> None:
    """Create a single vector index."""
    cosmos_options = {
        "kind": kind,
        "dimensions": dimensions,
        "similarity": similarity,
        **extra_params,
    }

    index_command = {
        "createIndexes": collection.name,
        "indexes": [
            {
                "name": name,
                "key": {vector_field: "cosmosSearch"},
                "cosmosSearchOptions": cosmos_options,
            }
        ],
    }
    collection.database.command(index_command)


def generate_embedding(azure_openai_client, query_text: str,
                       model_name: str) -> List[float]:
    """Generate a single embedding for the query text."""
    response = azure_openai_client.embeddings.create(
        input=[query_text],
        model=model_name
    )
    return response.data[0].embedding


def vector_search_with_index(collection, query_embedding: List[float],
                             vector_field: str,
                             top_k: int) -> List[Dict[str, Any]]:
    """Run vector search using the single active index and return results."""
    pipeline = [
        {
            "$search": {
                "cosmosSearch": {
                    "vector": query_embedding,
                    "path": vector_field,
                    "k": top_k
                }
            }
        },
        {
            "$project": {
                "document": "$$ROOT",
                "score": {"$meta": "searchScore"}
            }
        }
    ]

    results = list(collection.aggregate(pipeline))

    return results


def vector_search_with_retry(collection, query_embedding: List[float],
                             vector_field: str, top_k: int,
                             index_name_value: str) -> List[Dict[str, Any]]:
    """Retry vector search until results are available or retries are exhausted."""
    max_retries = 5
    retry_delay_seconds = 2

    for attempt in range(max_retries + 1):
        results = vector_search_with_index(
            collection, query_embedding, vector_field, top_k
        )
        if results:
            return results

        if attempt < max_retries:
            print(
                f"  No results for '{index_name_value}' yet. "
                f"Retrying in {retry_delay_seconds} seconds "
                f"({attempt + 1}/{max_retries})..."
            )
            time.sleep(retry_delay_seconds)

    print(
        f"  Search for '{index_name_value}' did not return results "
        f"after {max_retries} retries. Recording as failed."
    )
    return []


def main():
    print("=" * 70)
    print("  Compare All Algorithms — 9 Combinations")
    print("  (3 Algorithms × 3 Similarity Metrics)")
    print("=" * 70)

    config = get_compare_config()
    query_text = config["query_text"]
    top_k = config["top_k"]

    print(f"\n  Query:  \"{query_text}\"")
    print(f"  Top K:  {top_k}\n")

    mongo_client, azure_openai_client = get_clients_passwordless()

    try:
        database = mongo_client[config["database_name"]]

        # Drop collection for a clean comparison
        database.drop_collection("hotels")
        print("Dropped existing 'hotels' collection (if any)")

        # Create fresh collection and load data
        collection = database["hotels"]
        data = read_file_return_json(config["data_file"])
        documents = [doc for doc in data if config["vector_field"] in doc]
        print(f"Loaded {len(documents)} documents with embeddings")
        insert_data(collection, documents, config["batch_size"])

        # Generate ONE embedding for the query
        print("\nGenerating embedding for query...")
        query_embedding = generate_embedding(
            azure_openai_client, query_text, config["model_name"]
        )

        # Run all 9 searches sequentially (create→search→drop for each)
        print("Running 9 vector searches...\n")
        table_rows = []

        for algo_label, kind, extra_params in ALGORITHMS:
            for metric in METRICS:
                name = index_name(algo_label, metric)
                # Drop all vector indexes first
                drop_vector_indexes(collection, config["vector_field"])
                # Create this specific index
                create_vector_index(
                    collection, name, kind, config["vector_field"],
                    config["dimensions"], metric, extra_params
                )
                print(f"  Created index '{name}'")
                results = vector_search_with_retry(
                    collection, query_embedding, config["vector_field"], top_k, name
                )

                if not results:
                    table_rows.append([
                        algo_label,
                        metric,
                        "(failed)",
                        f"{0:.4f}",
                        "(failed)",
                        f"{0:.4f}",
                        f"{0:.4f}",
                    ])
                    continue

                top1_name = results[0].get("document", results[0]).get("HotelName", "Unknown") if len(results) > 0 else "(no results)"
                top1_score = results[0].get("score", 0) if len(results) > 0 else 0
                top2_name = results[1].get("document", results[1]).get("HotelName", "Unknown") if len(results) > 1 else "(no results)"
                top2_score = results[1].get("score", 0) if len(results) > 1 else 0

                table_rows.append([
                    algo_label,
                    metric,
                    top1_name,
                    f"{top1_score:.4f}",
                    top2_name,
                    f"{top2_score:.4f}",
                    f"{abs(top1_score - top2_score):.4f}",
                ])

        # Print comparison table
        headers = ["Algorithm", "Metric", "Top 1 Result", "Score",
                   "Top 2 Result", "Score", "Diff"]
        print(tabulate(table_rows, headers=headers, tablefmt="grid"))

    finally:
        # Cleanup: drop the comparison collection
        try:
            database = mongo_client[config["database_name"]]
            database.drop_collection("hotels")
            print("\nCleanup: dropped collection 'hotels'")
        except Exception as e:
            print(f"Cleanup warning: {e}")
        mongo_client.close()


if __name__ == "__main__":
    main()

File: ai/select-algorithm-go/src/compare_all.go
package main

import (
	"context"
	"fmt"
	"math"
	"strconv"
	"strings"
	"time"

	"github.com/openai/openai-go/v3"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
)

// CompareResult holds the result of a single algorithm+metric search
type CompareResult struct {
	Algorithm string
	Metric    string
	Results   []SearchResult
	Top1Name  string
	Top1Score float64
	Top2Name  string
	Top2Score float64
	Error     error
}

// indexSpec defines one of the 9 combinations
type indexSpec struct {
	Algorithm string
	Kind      string
	Metric    string
	IndexName string
	Options   bson.D
}

// RunCompareAll executes all 9 algorithm×metric combinations on a single collection
func RunCompareAll(ctx context.Context, config *Config, dbClient *mongo.Client, aiClient openai.Client) error {
	queryText := getEnvOrDefault("QUERY_TEXT", "luxury hotel near the beach")
	topK, _ := strconv.Atoi(getEnvOrDefault("TOP_K", "5"))

	fmt.Println("\n" + strings.Repeat("=", 70))
	fmt.Println("  COMPARE ALL: 3 Algorithms × 3 Similarity Metrics (9 combinations)")
	fmt.Println(strings.Repeat("=", 70))
	fmt.Printf("Query:  %q\n", queryText)
	fmt.Printf("Top-K:  %d\n", topK)

	// 1. Drop collection for clean comparison, then load data
	database := dbClient.Database(config.DatabaseName)
	collection := database.Collection("hotels")

	// Drop existing collection for a clean comparison
	if err := collection.Drop(ctx); err != nil {
		fmt.Printf("Note: could not drop collection (may not exist): %v\n", err)
	} else {
		fmt.Println("Dropped existing 'hotels' collection")
	}

	// Ensure cleanup on exit
	defer func() {
		fmt.Println("\nCleanup: dropping comparison collection...")
		if dropErr := collection.Drop(ctx); dropErr != nil {
			fmt.Printf("Cleanup warning: %v\n", dropErr)
		} else {
			fmt.Println("Cleanup: dropped collection 'hotels'")
		}
	}()

	fmt.Printf("\nLoading data from %s...\n", config.DataFile)
	data, err := ReadFileReturnJSON(config.DataFile)
	if err != nil {
		return fmt.Errorf("failed to load data: %v", err)
	}

	documentsWithEmbeddings := FilterDocumentsWithEmbeddings(data, config.VectorField)
	if len(documentsWithEmbeddings) == 0 {
		return fmt.Errorf("no documents found with embeddings in field '%s'", config.VectorField)
	}
	fmt.Printf("Loaded %d documents with embeddings\n", len(documentsWithEmbeddings))

	stats, err := PrepareCollection(ctx, collection, documentsWithEmbeddings, config.BatchSize)
	if err != nil {
		return err
	}
	fmt.Printf("Insertion completed: %d inserted, %d failed\n", stats.Inserted, stats.Failed)

	// 2. Generate ONE embedding for the query (reused for all 9 searches)
	fmt.Printf("\nGenerating embedding for query: %q\n", queryText)
	queryEmbedding, err := GenerateEmbedding(ctx, aiClient, queryText, config.ModelName)
	if err != nil {
		return fmt.Errorf("failed to generate query embedding: %v", err)
	}
	fmt.Printf("Embedding generated (%d dimensions)\n", len(queryEmbedding))

	// 3. Define all 9 index specs
	metrics := []string{"COS", "L2", "IP"}
	specs := buildIndexSpecs(config.VectorField, config.Dimensions, metrics)

	// 4. Create→search→drop each index sequentially (DocumentDB only allows one vector index per field)
	fmt.Printf("\nRunning %d vector index comparisons (create→search→drop)...\n", len(specs))
	var results []CompareResult
	successfulComparisons := 0
	failedComparisons := 0

	for _, spec := range specs {
		// Drop all existing vector indexes on this field
		DropVectorIndexes(ctx, collection, config.VectorField)

		// Create this specific index with retry (drop may still be in progress)
		var createErr error
		for attempt := 0; attempt < 3; attempt++ {
			if attempt > 0 {
				time.Sleep(3 * time.Second)
			}
			createErr = createNamedVectorIndex(ctx, collection, config.VectorField, spec)
			if createErr == nil {
				break
			}
		}
		if createErr != nil {
			results = append(results, CompareResult{
				Algorithm: spec.Algorithm,
				Metric:    spec.Metric,
				Error:     createErr,
			})
			failedComparisons++
			fmt.Printf("  ⚠ %s: %v\n", spec.IndexName, createErr)
			continue
		}
		fmt.Printf("  ✓ %s created\n", spec.IndexName)

		// Search using simple cosmosSearch with bounded retry for index readiness.
		searchResults, searchErr := runVectorSearchWithRetry(ctx, collection, queryEmbedding, config.VectorField, topK)

		top1Name, top1Score := extractResult(searchResults, 0)
		top2Name, top2Score := extractResult(searchResults, 1)

		cr := CompareResult{
			Algorithm: spec.Algorithm,
			Metric:    spec.Metric,
			Results:   searchResults,
			Top1Name:  top1Name,
			Top1Score: top1Score,
			Top2Name:  top2Name,
			Top2Score: top2Score,
			Error:     searchErr,
		}
		results = append(results, cr)
		if searchErr != nil {
			failedComparisons++
		} else {
			successfulComparisons++
		}
	}

	// 6. Print comparison table
	fmt.Println()
	printComparisonTable(results)
	fmt.Printf("\nSummary: %d succeeded, %d failed\n", successfulComparisons, failedComparisons)
	if successfulComparisons == 0 {
		return fmt.Errorf("all %d comparisons failed", failedComparisons)
	}

	return nil
}

func runVectorSearchWithRetry(ctx context.Context, collection *mongo.Collection, queryEmbedding []float64, vectorField string, topK int) ([]SearchResult, error) {
	const maxAttempts = 6
	const retryDelay = 2 * time.Second

	var searchResults []SearchResult
	var searchErr error

	for attempt := 1; attempt <= maxAttempts; attempt++ {
		searchResults, searchErr = vectorSearchSimple(ctx, collection, queryEmbedding, vectorField, topK)
		if searchErr == nil {
			if len(searchResults) > 0 {
				return searchResults, nil
			}
			searchErr = fmt.Errorf("search returned no results")
		}

		if attempt < maxAttempts {
			time.Sleep(retryDelay)
		}
	}

	return searchResults, searchErr
}

// buildIndexSpecs creates the 9 index specifications
func buildIndexSpecs(vectorField string, dimensions int, metrics []string) []indexSpec {
	var specs []indexSpec

	type algoConfig struct {
		name    string
		kind    string
		options bson.D
	}

	algos := []algoConfig{
		{"IVF", "vector-ivf", bson.D{{"numLists", 1}}},
		{"HNSW", "vector-hnsw", bson.D{{"m", 16}, {"efConstruction", 64}}},
		{"DiskANN", "vector-diskann", bson.D{{"maxDegree", 32}, {"lBuild", 50}}},
	}

	for _, algo := range algos {
		for _, metric := range metrics {
			metricLower := strings.ToLower(metric)
			opts := bson.D{
				{"kind", algo.kind},
				{"dimensions", dimensions},
				{"similarity", metric},
			}
			for _, o := range algo.options {
				opts = append(opts, o)
			}

			specs = append(specs, indexSpec{
				Algorithm: algo.name,
				Kind:      algo.kind,
				Metric:    metric,
				IndexName: fmt.Sprintf("vector_%s_%s", strings.ToLower(algo.name), metricLower),
				Options:   opts,
			})
		}
	}

	return specs
}

// createNamedVectorIndex creates a single named vector index
func createNamedVectorIndex(ctx context.Context, collection *mongo.Collection, vectorField string, spec indexSpec) error {
	indexCommand := bson.D{
		{"createIndexes", collection.Name()},
		{"indexes", []bson.D{
			{
				{"name", spec.IndexName},
				{"key", bson.D{
					{vectorField, "cosmosSearch"},
				}},
				{"cosmosSearchOptions", spec.Options},
			},
		}},
	}

	var result bson.M
	err := collection.Database().RunCommand(ctx, indexCommand).Decode(&result)
	if err != nil {
		if strings.Contains(err.Error(), "already exists") || strings.Contains(err.Error(), "IndexAlreadyExists") {
			return nil
		}
		return err
	}
	return nil
}

// vectorSearchSimple performs a vector search using the active vector index
func vectorSearchSimple(ctx context.Context, collection *mongo.Collection, embedding []float64, vectorField string, topK int) ([]SearchResult, error) {
	pipeline := []bson.M{
		{
			"$search": bson.M{
				"cosmosSearch": bson.M{
					"vector": embedding,
					"path":   vectorField,
					"k":      topK,
				},
			},
		},
		{
			"$project": bson.M{
				"document": "$$ROOT",
				"score":    bson.M{"$meta": "searchScore"},
			},
		},
	}

	cursor, err := collection.Aggregate(ctx, pipeline)
	if err != nil {
		return nil, err
	}
	defer cursor.Close(ctx)

	var results []SearchResult
	for cursor.Next(ctx) {
		var result SearchResult
		if err := cursor.Decode(&result); err != nil {
			continue
		}
		results = append(results, result)
	}

	if err := cursor.Err(); err != nil {
		return nil, err
	}

	return results, nil
}

// extractResult returns the name and score of the result at the given index
func extractResult(results []SearchResult, idx int) (string, float64) {
	if idx >= len(results) {
		return "(no results)", 0
	}
	doc := results[idx].Document.(bson.D)
	var name string
	for _, elem := range doc {
		if elem.Key == "HotelName" {
			name = fmt.Sprintf("%v", elem.Value)
			break
		}
	}
	if name == "" {
		name = "Unknown"
	}
	return name, results[idx].Score
}

// printComparisonTable outputs a formatted table of results
func printComparisonTable(results []CompareResult) {
	fmt.Println("┌──────────┬────────┬────────────────────────────┬────────┬────────────────────────────┬────────┬───────┐")
	fmt.Printf("│ %-8s │ %-6s │ %-26s │ %-6s │ %-26s │ %-6s │ %-5s │\n",
		"Algorithm", "Metric", "Top 1 Result", "Score", "Top 2 Result", "Score", "Diff")
	fmt.Println("├──────────┼────────┼────────────────────────────┼────────┼────────────────────────────┼────────┼───────┤")

	for _, r := range results {
		if r.Error != nil {
			fmt.Printf("│ %-8s │ %-6s │ %-26s │ %-6s │ %-26s │ %-6s │ %-5s │\n",
				r.Algorithm, r.Metric, "ERROR", "-", "-", "-", "-")
			continue
		}

		top1 := r.Top1Name
		if len(top1) > 26 {
			top1 = top1[:26]
		}
		top2 := r.Top2Name
		if len(top2) > 26 {
			top2 = top2[:26]
		}
		diff := math.Abs(r.Top1Score - r.Top2Score)

		fmt.Printf("│ %-8s │ %-6s │ %-26s │ %6.4f │ %-26s │ %6.4f │%6.4f │\n",
			r.Algorithm, r.Metric, top1, r.Top1Score, top2, r.Top2Score, diff)
	}

	fmt.Println("└──────────┴────────┴────────────────────────────┴────────┴────────────────────────────┴────────┴───────┘")
}
